File: watch/JarvisWatchApp/SocketManager.swift
import Foundation
import Combine
import WatchKit

class SocketManager: ObservableObject {
    @Published var isConnected = false
    @Published var currentStatus = "Disconnected"
    @Published var latestAnswer: String?
    @Published var isReasoning = false
    
    private var webSocketTask: URLSessionWebSocketTask?
    
    // Replace with your Mac's local IP Address
    private let macIP = "192.168.1.100" 
    private let deviceToken = "JARVIS-WATCH-001"
    
    func connect() {
        guard let url = URL(string: "ws://\\(macIP):8000/api/mobile/ws/\\(deviceToken)") else { return }
        
        let session = URLSession(configuration: .default)
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()
        
        receiveMessage()
    }
    
    func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        isConnected = false
        currentStatus = "Disconnected"
    }
    
    func sendCommand(_ command: String) {
        let payload: [String: Any] = ["type": "command", "text": command]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let jsonString = String(data: data, encoding: .utf8) else { return }
        
        let message = URLSessionWebSocketTask.Message.string(jsonString)
        webSocketTask?.send(message) { error in
            if let error = error {
                print("Watch send error: \\(error)")
            } else {
                DispatchQueue.main.async {
                    self.isReasoning = true
                    self.latestAnswer = nil
                    WKInterfaceDevice.current().play(.success)
                }
            }
        }
    }
    
    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .failure(let error):
                print("Watch socket error: \\(error)")
                DispatchQueue.main.async {
                    self?.isConnected = false
                    self?.currentStatus = "Error connecting"
                }
                
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.handleMessage(text)
                default: break
                }
                self?.receiveMessage()
            }
        }
    }
    
    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        
        DispatchQueue.main.async {
            let type = json["type"] as? String ?? ""
            
            if type == "system" {
                self.isConnected = true
                self.currentStatus = "Connected to Mac"
                WKInterfaceDevice.current().play(.click)
            } 
            else if type == "state_stream" {
                if let payload = json["data"] as? [String: Any] {
                    if let answer = payload["answer"] as? String {
                        self.latestAnswer = answer
                        self.isReasoning = false
                        WKInterfaceDevice.current().play(.notification)
                    }
                }
            }
            else if type == "ack" {
                self.currentStatus = "Executing..."
            }
        }
    }
}

File: mobile/src/services/SocketManager.js
// SocketManager.js
export default class SocketManager {
  constructor(ip, token, onStateChange, onMessage) {
    this.ip = ip;
    this.token = token;
    this.onStateChange = onStateChange;
    this.onMessage = onMessage;
    this.ws = null;
    this.reconnectTimer = null;
  }

  connect() {
    if (this.ws) return;
    this.onStateChange('connecting');

    const url = `ws://${this.ip}:8000/api/mobile/ws/${this.token}`;
    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      this.onStateChange('connected');
      if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    };

    this.ws.onmessage = (e) => {
      try {
        const data = JSON.parse(e.data);
        if (this.onMessage) this.onMessage(data);
      } catch (err) {
        console.warn('Failed to parse WS msg:', err);
      }
    };

    this.ws.onclose = () => {
      this.ws = null;
      this.onStateChange('disconnected');
      this.reconnectTimer = setTimeout(() => this.connect(), 3000);
    };

    this.ws.onerror = (e) => {
      console.warn('WS Error:', e.message);
      this.ws?.close();
    };
  }

  sendCommand(text) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ type: 'command', text }));
    }
  }

  disconnect() {
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}
